app/src/org/commcare/utils/SigningUtil.java

```diff
@@ -50,12 +59,53 @@
         return new Pair<>(downloadLink, signatureBytes);
     }
 
+    /**
+     * Given a base64 encoded URL, decode the URL, and return first line read
+     * from accessing that URL
+     */
+    public static String convertEncodedUrlToPayload(String baseEncodedUrl)
+            throws IOException, Base64DecoderException {
+        return readURL(decodeUrl(baseEncodedUrl));
+    }
+
+    protected static String decodeUrl(String baseEncodedUrl)
+            throws Base64DecoderException, UnsupportedEncodingException {
+        String decodedUrl;
+        if (baseEncodedUrl.startsWith("http://") || baseEncodedUrl.startsWith("https://")) {
+            // for backwards compatibility, accept non-base64 encoded URLS
+            // once all users have migrated to the new format
+            // (info available on HQ?) we can remove this branch
+            decodedUrl = baseEncodedUrl;
+        } else {
+            decodedUrl = new String(Base64.decode(baseEncodedUrl), "UTF-8");
+        }
+        assertWhitelistedUrlHost(decodedUrl);
+        return decodedUrl;
+    }
 
     /**
-     * Given a URL, return the text at that location as a String
+     * Very basic method to prevent spoofed SMSs from making CommCare hit malicious URLs.
      */
-    public static String convertUrlToPayload(String url) throws IOException {
-        return readURL(url);
+    private static void assertWhitelistedUrlHost(String urlString) {
+        URL url;
+        try {
+            url = new URL(urlString);
+        } catch (MalformedURLException e) {
+            throw new RuntimeException(urlString + " is not a valid URL.");
+        }
+
+        String host = url.getHost();
+        boolean isURLWhitelisted = false;
+        for (Pattern urlPattern : WHITELISTED_URL_HOSTS_REGEX) {
+            if (urlPattern.matcher(host).find()) {
+                isURLWhitelisted = true;
+                break;
+            }
+        }
+
+        if (!isURLWhitelisted) {
+            throw new RuntimeException(url + " is not an approved URL.");
```

Comment:
The sms processing async task is wrapped in a broad try/catch to prevent this. I've updated it to throw a specific RuntimeException subclass, but I've left the umbrella try/catch in place. It still reports a generic 'sms install failed' toast message to the user if this comes up.